Automattic/wp-calypso
# Async load reducers and persist state in parts

## Already merged in separate PRs

**Adding a new reducer**
Patching `combineReducers` and adding a store enhancer helps us to add a new reducer to any point in the tree:
```js
store.addReducer( [ 'extensions', 'woocommerce' ], wooReducer );
```

**Serializing any subtree into separate IndexedDB rows**
Decorate your reducer with `withStorageKey`:
```js
export default withStorageKey( 'woocommerce', wooReducer );
```
and pass it to `combineReducers` as any other reducer.

When serializing, the state will be stored into multiple IndexedDB rows:
```
redux-state-123-> root state without woo
redux-state-123:woocommerce -> the woo subtree alone
```

## Proposed in this PR

**Load and deserialize state at the right moment**
Load the state for a given `storageKey` from IndexedDB just in time.

If the `storageKey` is a part of the initial reducer, do it during store initialization. Implemented using the new `reducer.getStorageKeys` method and with multiple IndexedDB loads in `getInitialStoredState`.

The `getStateFromLocalStorage` is now used to load state pieces of various shapes. That's why it newly gets the specific reducer as a new explicit parameter. The reducer is used to deserialize the stored state.

If a reducer is added dynamically, the stored state is loaded right after the reducer is added. That's what the new `addReducerToStore` function does.

**Add reducers in extension (section) entrypoint**
Migrate the extensions' sections to this new system. The section initialization function get a new parameter (`addReducer`) that can be used to add a new reducer when the section is being loaded.

The extension reducer are removed from the initial reducer and the webpack loader that extracted reducers from sections is removed, too.

## Testing instructions
Let's test on the Zoninator extension, for example:
1. Go to `calypso.localhost:3000/?flags=-calypsoify/plugins`, disabling the redirect of Plugins UI to wp-admin.
2. Install the Zoninator extension to your Jetpack/Atomic site.
3. Go to `/extensions/zoninator/your.blog` and add a zone.
4. Reload the page.

<img width="551" alt="screenshot 2018-12-03 at 19 20 01" src="https://user-images.githubusercontent.com/664258/49396221-e78ff280-f738-11e8-8190-d5a001a7eddc.png">

You should see that the zone list is displayed immediately, populated from persisted state, rather than showing an intermediate placeholder while a REST request is in flight.

**Verifying the state's inner workings**
Verify that the state got persisted into two separate keys:

<img width="432" alt="screenshot 2018-12-03 at 19 21 28" src="https://user-images.githubusercontent.com/664258/49396341-2aea6100-f739-11e8-8d5f-397847b8be55.png">

In console, looking at `state.extensions`, verify that the state for `state.extensions.zoninator` doesn't exist at all before you go to an `/extensions/zoninator` route.
